cpp/src/text/edit_distance.cu

```diff
@@ -108,16 +94,179 @@ struct calculate_compute_buffer_fn {
+           static_cast<int64_t>(count);
   }
 };
 
-}  // namespace
-
 /**
- * @copydoc nvtext::edit_distance
+ * @brief Processes the 2 given strings by computing the matrix values
+ * comparing each character of both strings
+ *
+ * The full matrix is computed from top-left to bottom-right with the edit-distance
+ * as the final value. Only 3 diagonals are needed in memory at any iteration.
+ *
+ *     -  S  E  C  O  N  D
+ *  -  0  1  2  3  4  5  6
+ *  F  1  =  x  y  z  ↘  .
+ *  I  2  x  y  z  ↘  .  .
+ *  R  3  y  z  ↘  .  .  .
+ *  S  4  z  ↘  .  .  .  .
+ *  T  5  →  .  .  .  .  X
+ *
+ * Computing a diagonal only requires the previous 2 and each value
+ * can be computed independently/parallel a tile at a time.
+ *
+ * The diagonal approach inspired by:
+ * https://ashvardanian.com/posts/stringwars-on-gpus/#dynamic-programming-and-levenshtein-evaluation-order
  */
+template <typename Iterator, int32_t tile_size = cudf::detail::warp_size>
+__device__ void compute_distance(Iterator input1,
+                                 cudf::size_type length1,
+                                 Iterator input2,
+                                 cudf::size_type length2,
+                                 cudf::size_type* d_buffer,
+                                 cudf::size_type* d_result)
+{
+  namespace cg        = cooperative_groups;
+  auto const block    = cg::this_thread_block();
+  auto const tile     = cg::tiled_partition<tile_size>(block);
+  auto const lane_idx = tile.thread_rank();
+
+  // shortcut if one of the strings is empty
+  // (null rows are mapped here as well)
+  if (length1 == 0 || length2 == 0) {
+    if (lane_idx == 0) { *d_result = length1 == 0 ? length2 : length1; }
+    return;
+  }
+
+  // setup the 3 working vectors for this string
+  auto v0 = d_buffer;
+  auto v1 = v0 + length1 + 2;
+  auto v2 = v1 + length1 + 2;
+  if (lane_idx == 0) {
+    v0[0] = 0;  // first diagonal
+    v1[0] = 1;  // second diagonal
+    v1[1] = 1;
+  }
+
+  // utility for navigating the diagonal of the matrix of characters for the 2 strings
+  auto next_itr = [](Iterator sitr, cudf::size_type length, Iterator itr, cudf::size_type offset) {
+    if constexpr (cuda::std::is_pointer_v<Iterator>) {
+      itr = sitr - offset;  // ASCII iterator
+    } else {
+      auto const pos = sitr.position() - offset;  // minimizes character counting
+      itr += (pos >= 0) && (pos < length) ? (pos - itr.position()) : 0;
+    }
+    return itr;
+  };
+
+  // top-left of the matrix
+  // includes the diagonal one after the max(length1,length2) diagonal
+  for (auto idx = 0; idx < length1; ++idx, ++input1) {
+    auto const n = idx + 2;      // diagonal length
+    auto const a = n > length1;  // extra diagonal adjust
+
+    auto jdx = static_cast<cudf::size_type>(lane_idx);
+    auto it1 = input1;
+    auto it2 = input2;
+
+    auto tile_count = cudf::util::div_rounding_up_safe(n + 1, tile_size);
+    while (tile_count--) {
+      auto const offset = (jdx - 1);
```

Comment:
Is this `-1` because we want to skip the 0th element that is initialized manually above?

Reply:
It's for addressing the previous character in the vertical direction or the next character in the horizontal direction as the diagonal is processed.